doc/sfp-cmis/cmis-init.md

```diff
@@ -0,0 +1,387 @@
+post its module information onto the **STATE_DB**
+- When a CMIS transceiver is attached, the **show interfaces transceiver eeprom**
+should display the advertised applications of the transceiver.
+- When a CMIS transceiver is attached, the **sfputil** should be capable of reporting
+the errors of the CMIS application initialization.
+- When a CMIS transceiver is detected, the **pmon#xcvrd** should automatically activate
+the appropriate application as per the current port mode.
+- Only the default application mode will be supported for now, we'll deal with the dynamic
+application update later when the synchronization mechanism between **syncd** and **pmon#xcvrd**
+is in place and ready.
+
+# Design
+
+## sonic-platform-daemons/sonic-xcvrd (modified)
+
+The transceiver daemon will be enhanced as below:
+
+- The **application_advertisement** in the **TRANSCEIVER_INFO** table is now updated
+to the json format:  
+
+**Original:**  
+```
+  "TRANSCEIVER_INFO|Ethernet0": {
+    "type": "hash",
+    "value": {
+      "application_advertisement": "400GAUI-8 C2M (Annex 120E) - 400GBASE-DR4 (Cl 124)\n        100GAUI-2 C2M (Annex 135G) - 100G-FR/100GBASE-FR1 (Cl 140)",
+      ...... omitted ......
+    }
+  },
+```
+
+**Modified:**  
+```
+  "TRANSCEIVER_INFO|Ethernet0": {
+    "type": "hash",
+    "value": {
+      "application_advertisement": "{1: {'host_electrical_interface_id': '400GAUI-8 C2M (Annex 120E)', 'module_media_interface_id': '400GBASE-DR4 (Cl 124)', 'host_lane_count': 8, 'media_lane_count': 4, 'host_lane_assignment_options': 1, 'media_lane_assignment_options': None}, 2: {'host_electrical_interface_id': '100GAUI-2 C2M (Annex 135G)', 'module_media_interface_id': '100G-FR/100GBASE-FR1 (Cl 140)', 'host_lane_count': 2, 'media_lane_count': 1, 'host_lane_assignment_options': 85, 'media_lane_assignment_options': None}}",
+      ...... omitted ......
+    }
+  },
+  ...... omitted ......
+```  
+
+- Add **CmisManagerTask** for the state-based CMIS application initialization to support
+multiple CMIS transceivers in one single thread.
+
+  ![](images/001.png)
+
+- Enhance **xcvrd_utilities/port_mapping.py** with **APPL_DB** and **STATE_DB** support
+  - **port_mapping.subscribe_port_config_change()** only listens to **CONFIG_DB** by default.
+  - **port_mapping.subscribe_port_config_change(['APPL_DB', 'STATE_DB'])** listens to both **APPL_DB** and **STATE_DB**.
+  - The port table name associated with the database  
+
+    | Database  |      Table Name        |
+    |:--------- |:---------------------- |
+    | APPL_DB   | PORT_TABLE             |
+    | CONFIG_DB | PORT                   |
+    | STATE_DB  | TRANSCEIVER_INFO       |
+
+- Introduce the following additional events into **xcvrd_utilities/port_mapping.py**  
+  - **PortChangeEvent.PORT_SET**  
+    The event callbacks for database **SET** operations.
+  - **PortChangeEvent.PORT_DEL**  
+    The event callbacks for database **DEL** operations.
```

Comment:
Does it need to handle this case:  NO port add/del but the cable is replaced by a new one?

Reply:
Yes, that's the reason why PORT_SET/PORT_DEL is introduced in the code PR, the callbacks will be invoked upon all the SET/DEL events to restart the CMIS init if applicable, while the original implementation of portmap only trigger the callback upon port add/remove compared to the logical port list maintained in the thread-specific memory

Reply:
understood, it can sense the cable replace event via monitoring STATE DB.